app/code/Magento/CatalogInventory/Model/StockStateProvider.php

```diff
@@ -106,6 +106,10 @@ public function checkQuoteItemQty(StockItemInterface $stockItem, $qty, $summaryQ
     {
         $result = $this->objectFactory->create();
         $result->setHasError(false);
+        
+        $product = $this->productFactory->create();
+        $product->load($stockItem->getProductId());
+        $stockItem->setProduct($product);
```

Comment:
I do believe that the code works ;)

The problem is that this method is not part of `StockItemInterface`.

Lame way is just add `setProduct` to mock in failed test. The downside is that the code we have do not follow contract defined by `StockItemInterface`.

Better way is achieve the same correct behavior but use only defined contracts and do not rely on existing implementation. Why wrong product instance appear in `StockItem` at all? Of course, it may appear, that correct fix require much more efforts in such case breaking contract with `setProduct` call is lesser evil than having bug.